src/main/java/org/humancellatlas/ingest/core/service/strategy/impl/ProcessCrudStrategy.java

```diff
@@ -48,4 +54,32 @@ public Process findOriginalByUuid(String uuid) {
     public Collection<Process> findAllBySubmissionEnvelope(SubmissionEnvelope submissionEnvelope) {
         return processRepository.findAllBySubmissionEnvelope(submissionEnvelope);
     }
+
+    @Override
+    public void removeLinksToDocument(Process document) {
+        messageRouter.routeStateTrackingDeleteMessageFor(document);
+        fileRepository.findByInputToProcessesContains(document).forEach(file -> {
+            file.getInputToProcesses().remove(document);
+            fileRepository.save(file);
+        });
+        fileRepository.findByDerivedByProcessesContains(document).forEach(file -> {
+            file.getDerivedByProcesses().remove(document);
+            fileRepository.save(file);
+        });
+        biomaterialRepository.findByInputToProcessesContains(document).forEach(biomaterial -> {
+            biomaterial.getInputToProcesses().remove(document);
+            biomaterialRepository.save(biomaterial);
+        });
+        biomaterialRepository.findByDerivedByProcessesContains(document).forEach(biomaterial -> {
+            biomaterial.getDerivedByProcesses().remove(document);
+            biomaterialRepository.save(biomaterial);
+        });
+
+    }
```

Comment:
We can reuse this for optimising delete, good job on this!

Reply:
And we can probably get it even better: https://github.com/ebi-ait/ingest-core/pull/119#discussion_r840658856